### PR TITLE
Proposal: Container Widgets as Context Managers

### PR DESCRIPTION
**Pull request by HalfWhitt**
Take a gander at this and see if it looks like a direction you'd be interested in.

Inspired by the context-manager-based syntax shown in #118, and using techniques similar to those used by [PyMC](https://github.com/pymc-devs/pymc) as helpfully explained in [this StackOverflow question](https://stackoverflow.com/a/51849809/12975140) (credit where credit is due!), I've put together a quick proof of concept.

With these changes, this code from the [Slightly Less Toy Example](https://toga.readthedocs.io/en/stable/tutorial/tutorial-1.html) tutorial:

```python
c_input = toga.TextInput(readonly=True)
f_input = toga.TextInput()

c_label = toga.Label("Celsius", style=Pack(text_align=LEFT))
f_label = toga.Label("Fahrenheit", style=Pack(text_align=LEFT))
join_label = toga.Label("is equivalent to", style=Pack(text_align=RIGHT))

button = toga.Button("Calculate", on_press=calculate)

f_box.add(f_input)
f_box.add(f_label)

c_box.add(join_label)
c_box.add(c_input)
c_box.add(c_label)

box.add(f_box)
box.add(c_box)
box.add(button)
```

can become this:

```python
with toga.Box() as box:
    with toga.Box() as f_box:
        f_input = toga.TextInput()
        f_label = toga.Label("Fahrenheit", style=Pack(text_align=LEFT))

    with toga.Box() as c_box:
        join_label = toga.Label("is equivalent to", style=Pack(text_align=RIGHT))
        c_input = toga.TextInput(readonly=True)
        c_label = toga.Label("Celsius", style=Pack(text_align=LEFT))

    button = toga.Button("Calculate", on_press=calculate)
```

(The variable assignments aren't functionally necessary for what's shown here, but would still be needed if setting the styles afterward like in the tutorial.)

Personally I find it much easier to visually grok the structure in this form — much like in a declarative syntax (which keeps coming up), but in this case directly in Python, with no need to parse a DSL or process a data file.

I don't know if the thread-local storage is overkill or not... I'm not sure if people are likely to create widgets in multiple threads, or if that's even guaranteed to work in Toga as it is. And obviously decisions would have to be made about how to implement it for other widgets — particularly ScrollContainer, which would need a way to be given the flex values alongside the widgets. No sense going down all the roads before checking in though. Thoughts?

## PR Checklist:

- [ ] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by freakboy3742** [comment]
I've seen this sort of pattern many times in the past - it usually comes up in the context of HTML templating. I'm not a huge fan of it - mostly because it feels like an abuse of context managers, but also because the API/descriptive pattern it's creating is already possible:

```python
box = toga.Box(children=[
    toga.Box(children=[
        toga.TextInput(),
        toga.Label("Fahrenheit", style=Pack(text_align=LEFT)),
    ],
    toga.Box(children=[
        toga.Label("is equivalent to", style=Pack(text_align=RIGHT)),
        toga.TextInput(readonly=True),
        toga.Label("Celsius", style=Pack(text_align=LEFT)),
    ],
    toga.Button("Calculate", on_press=calculate),
])
```
(And, to circumvent the next critique: if you need to reference one of these objects later, you can use `:=` to assign them, or use ID-based widget lookup).

> I don't know if the thread-local storage is overkill or not... I'm not sure if people are likely to create widgets in multiple threads, or if that's even guaranteed to work in Toga as it is. And obviously decisions would have to be made about how to implement it for other widgets — particularly ScrollContainer, which would need a way to be given the flex values alongside the widgets. No sense going down all the roads before checking in though. Thoughts?

In short - I'm not a fan. In addition to the "you can already do this" criticism, the specific implementation is based on threadlocals and an implied association between a widget and its context. This sort of implicit assignment seems counter to the general sprit of Python (and the "explicit is better than implicit" zen koan). 

On that basis, I'm going to say "thanks, but no thanks", and close this PR. Definitely thanks for the suggestion though - it's always worth exploring new ideas, even if they don't go anywhere; and having working code demonstrates you're not just doing a pie-in-the-sky "wouldn't it be great if ..." speculation.

**Comment by HalfWhitt** [comment]
Fair enough! I do tend to lean more toward the abstracting-things-into-tidiness direction — probably why I'm such a fan of descriptors — but I get that sometimes too much magic is too much.
